### ISSUE TITLE
Implementation of stateful built-ins

### ISSUE THREAD

**Issue by crb233** [crb233/unarian#5]
It would be nice to have I/O built-ins `?` and `!` added to the official specification (no longer just for debugging). Another useful built-in would be a 'coin flip' `%` which fails with probability 1/2 and acts like the identity function otherwise. However, these stateful operations don't seem to play well with Unarian.

The language currently has a nice property where `A { B | C }` is semantically equivalent to `{ A B | A C }` (although an analogous property doesn't hold: `{ A | B } C` is not the same as `{ A C | B C }`). So I would want `? { - | + }` to be equivalent to `{ ? - | ? + }`. However, the naive implementation of `?` which asks for new input every time it's encountered can break this property, since the second expression sometimes asks for input twice. The alternative is to store a buffer of inputs so that every nth call to `?` produces the same result. Unfortunately, solving the same problem for `!` creates a new problem where we can't print out anything until the program exits (since we could be on a failing execution path and will overwrite previously printed values with new ones). This is obviously undesirable if we want to print as we go or run programs that might never terminate.

At the moment, I'm leaning toward throwing out the property `A { B | C }` = `{ A B | A C }` in favor of having simple stateful operations.

**Comment by DivergentClouds** [comment]
I think it'd be nice to have output without input as that wouldn't require modifying the state. I have a similar argument against the coin flip operation. I am in favor of the output builtin just printing out the current input as a number and otherwise acting as the identity. I quite like the property `A { B | C }` = `{ A B | A C }`, I think it makes the language feel cleaner.

**Comment by crb233** [comment]
Even `!` breaks this property, in a weaker sense. It doesn't affect the program execution, but it does affect what the user sees. `! { - | + }` always prints exactly once, but `{ ! - | ! + }` may print twice.

**Comment by DivergentClouds** [comment]
I feel like since it doesn't effect internal state it's more in line with the language's design philosophy.

**Comment by DivergentClouds** [comment]
regarding `?`, I think the best way to handle that might be to allow specifying a starting number for the program. That way input is always given exactly once. (no input being given could be equivelant to 0 being given.)

**Comment by DivergentClouds** [comment]
Looking back on this, a thought I had to the problem with `!`  may be for an interpreter to note when the last `n` printed inputs have been from a failed execution path. What are your thoughts on that?

**Comment by crb233** [comment]
That's a really interesting idea, and thinking about this again led me to consider a fourth possibility too. In what cases is it actually useful for stateful functions like `!` and `?` to be undone, in the sense that we print a value or ask for user input and then later "retcon" that action and pretend it never happened? In Unarian terms, in what cases is it practically useful for such a function to be called within a failing branch of an alternation? Assuming there's no need for this behavior, the interpreter can require that stateful functions are never undone, i.e. once either `!` or `?` is encountered within some branch of an alternation, no other alternate branches are ever considered, even if the current one fails.

In some way, this solution is less elegant than yours, because functions `!` and `?` get special treatment that totally change the execution behavior of a program. On the other hand, it is more practical for users of programs with a lot of IO because each IO operation is guaranteed to be part of the "final" execution path of the program. In particular, for programs that run for a long time and print out many values as they go, the user can never be sure that all the values printed so far won't later be undone. Still, whenever a program is written with the guarantee that alternate branches containing `!` or `?` will never fail, then these two approaches are indistinguishable.

Overall, I am leaning toward your suggested approach now. It's elegant and should be fairly easy to implement. There's also a sense in which it maintains the left-distributive property `A { B | C }` = `{ A B | A C }` in the long run.

**Comment by DivergentClouds** [comment]
A problem with your proposed idea is that `!` doesn't really offer a benefit then, because if you are for instance, running a loop and wanting to print the value of each iteration, calling the next iteration would still be considered part of the same path, no? So it would guarantee that the rest of the entire loop is non-failing.   I could be misunderstanding, but ultimately, I think your idea is equivalent in usefulness to just printing out all values at the end of the program. I feel like I am missing something though, so feel free to correct me.

**Comment by crb233** [comment]
No, that's not quite right. Consider the function `main { ! { - main | } }`. On input $n$, it should print $n$, $n - 1$, ... $1$, $0$, and then exit. Observe that failure can only occur at the `-` and this can only happen within an alternation before any calls to `!` in the same branch. Therefore, failure will never occur in a branch after a value has been printed.

It's not exactly the same as printing out all values at the end of the program because values are printed in real time and once printed are never undone. This is useful especially for programs that run for a long time or don't terminate at all but still produce useful information as they go (e.g. listing Fibonacci numbers or primes).

Still, I am leaning toward your solution. The only major drawback I can see is how to effectively communicate to the user that the last $n$ values printed have been undone. Even more broadly, we may want to be able to hook the output of one program into the input of another through something like a pipe. If output can be undone, then we need a way to communicate that with the program receiving that output.

**Comment by DivergentClouds** [comment]
I understand now, I do quite like your solution actually, and the point about piping is excellent. What if, building off your solution, not only is the current path not allowed to fail, but it is not allowed to _potentially_ fail. I'm not quite sure how to prove that during parsing, though.

**Comment by crb233** [comment]
That would be great if possible, but to determine this exactly would require an algorithm that can decide for arbitrary expressions `A` and `B` whether for all possible inputs either `A` fails or `A B` doesn't fail. This appears undecidable in general, so any terminating implementation would have to reject some valid programs which it can't easily check.

**Comment by DivergentClouds** [comment]
right, that makes sense. I do still prefer your idea though. The interpreter could give an error if the assertion fails.

**Comment by DivergentClouds** [comment]
additionally, if the output could fail and the failure is communicated as per the idea I had, it could be hard for a human to tell at a glance what inputs are true, additionally, consider the program `main { !  { ! - | } - }`. If it is initialized with the number `0` then the output would have to be something like

```
0
0
[discard last 1 value(s)]
[discard last ? value(s)]
```

would `?` be 1 or 2? because 1 of the values was already discarded; either way I think it would appear ambiguous to a human reading the output.

**Comment by crb233** [comment]
Maybe any failure of a branch after printing  could be considered a runtime error, and interpreters / compilers are suggested but not required to warn about obvious instances of this error.

**Comment by DivergentClouds** [comment]
that sounds ideal to me

**Comment by DivergentClouds** [comment]
I think `?` should use the buffer of inputs approach described in the original post in this issue.

**Comment by DivergentClouds** [comment]
Oh, and I think the stack trace builtin `@` should not be subject to the restriction that `!` is, as the stack trace is exclusively for debug purposes.
